ext/libxml/ruby_xml_node.c

```diff
@@ -45,13 +45,18 @@ static void rxml_node_deregisterNode(xmlNodePtr xnode)
   /* Has the node been wrapped and exposed to Ruby? */
   if (xnode->_private)
   {
+    /* Does it belong to another libxml user? */
+    if (!OWNED(xnode))
```

Comment:
The hash table lookup should be fast.  So unless we have proof that this it is actually slow (i.e., a profile run showing it to be) I don't think this should be optimized now.  Let's go with the clean solution, and if it turns out to be a problem then like you said its easy to add quick check by setting a special value.